platform/utm_mgrs_utils.cpp

```diff
@@ -0,0 +1,567 @@
+const double E2 = E * E;
+const double E3 = E2 * E;
+const double E_P2 = E / (1 - E);
+const double R = 6378137.0;
+
+const double SQRT_E = sqrt(1 - E);
+const double _E = (1 - SQRT_E) / (1 + SQRT_E);
+const double _E2 = _E * _E;
+const double _E3 = _E2 * _E;
+const double _E4 = _E3 * _E;
+const double _E5 = _E4 * _E;
+
+const double M1 = (1.0 - E / 4.0 - 3.0 * E2 / 64.0 - 5.0 * E3 / 256.0);
+const double M2 = (3.0 * E / 8.0 + 3.0 * E2 / 32.0 + 45.0 * E3 / 1024.0);
+const double M3 = (15.0 * E2 / 256.0 + 45.0 * E3 / 1024.0);
+const double M4 = (35.0 * E3 / 3072.0);
+
+const double P2 = (3.0 / 2.0 * _E - 27.0 / 32.0 * _E3 + 269.0 / 512.0 * _E5);
+const double P3 = (21.0 / 16.0 * _E2 - 55.0 / 32.0 * _E4);
+const double P4 = (151.0 / 96.0 * _E3 - 417.0 / 128.0 * _E5);
+const double P5 = (1097.0 / 512.0 * _E4);
+
+const string ZONE_LETTERS = "CDEFGHJKLMNPQRSTUVWXX";
+double const DEGREES_TO_RADIANS = 0.017453292519943295;
+double const RADIANS_TO_DEGREES = 57.29577951308232;
+
+const int NUM_100K_SETS = 6;
+const int SET_ORIGIN_COLUMN_LETTERS[] = { 'A', 'J', 'S', 'A', 'J', 'S' };
+const int SET_ORIGIN_ROW_LETTERS[] = { 'A', 'F', 'A', 'F', 'A', 'F' };
+
+
+double degToRad(double deg)
+{
+  return deg * DEGREES_TO_RADIANS;
+}
+
+double radToDeg(double rad)
+{
+  return rad * RADIANS_TO_DEGREES;
+}
+
+// Convert lat,lon for WSG 84 ellipsoid to MGRS string.
+string FormatMGRS(double lat, double lon, int precision)
+{
+  if (precision > 5)
+    precision = 5;
+  else if (precision < 1)
+    precision = 1;
+
+  if (lat <= -80 || lat > 84)
+    return "Latitude limit exceeded";
+  if (lon <= -180 || lon > 180)
+    return "Longitude limit exceeded";
+
+  UTMPoint mgrsp = latlon_to_utm(lat, lon);
+
+  // Need to add this to set the right letter for the latitude.
+  mgrsp.zone_letter = latitude_to_zone_letter(lat);
+  return utm_to_mgrs_str(mgrsp, precision);
+}
+
+// Convert lat,lon for WSG 84 ellipsoid to UTM string.
+string FormatUTM(double lat, double lon)
+{
+  if (lat <= -80 || lat > 84)
+    return "Latitude limit exceeded";
+  if (lon <= -180 || lon > 180)
+    return "Longitude limit exceeded";
+
+  UTMPoint utm = latlon_to_utm(lat, lon);
+  return utm_to_str(utm);
+}
+
+// Main algorithm. Formulars sourcre: https://github.com/Turbo87/utm
+UTMPoint latlon_to_utm(double lat, double lon)
+{
+  double lat_rad = degToRad(lat);
+  double lat_sin = sin(lat_rad);
+  double lat_cos = cos(lat_rad);
+
+  double lat_tan = lat_sin / lat_cos;
+  double lat_tan2 = lat_tan * lat_tan;
+  double lat_tan4 = lat_tan2 * lat_tan2;
+
+  int zone_number = latlon_to_zone_number(lat, lon);
+  char zone_letter = (lat >= 0.0) ? 'N' : 'S';
+
+  double lon_rad = degToRad(lon);
+  double central_lon = zone_number_to_central_longitude(zone_number);
+  double central_lon_rad = degToRad(central_lon);
+
+  double n = R / sqrt(1 - E * lat_sin * lat_sin);
+  double c = E_P2 * lat_cos * lat_cos;
+
+  double a = lat_cos * mod_angle(lon_rad - central_lon_rad);
```

Comment:
All const variables should be marked as const.

Reply:
Fixed.